stream_pool_peer.go

```diff
@@ -0,0 +1,161 @@
+package dht
+
+import (
+	"context"
+	"sync"
+
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+	"go.opencensus.io/trace"
+	"golang.org/x/xerrors"
+)
+
+type peerStreamPool struct {
+	// Callback that returns a stream. The pool exists to minimize calls to this.
+	newStream func(context.Context) (*stream, error)
+
+	mu sync.Mutex
+	// Spare streams not in use
+	streams map[*stream]struct{}
+	// Sync machinery for goroutines waiting on streams.
+	waiters map[*streamWaiter]struct{}
+	// This is the number of new streams being created concurrently.
+	pending int
+
+	// This synchronizes all send-only messages to minimize unnecessary concurrent stream use.
+	sendMu sync.Mutex
+}
+
+// State for someone waiting on a stream.
+type streamWaiter struct {
+	s    *stream
+	err  error
+	done bool
+	// Unlocked when the above fields are ready.
+	ret sync.Mutex
+}
+
+func (me *peerStreamPool) getStream(ctx context.Context) (*stream, error) {
+	ctx, span := trace.StartSpan(ctx, "get stream")
```

Comment:
Can you add some empty lines between logical sections of the code? This is is too dense.

Reply:
Fixed